docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,368 @@
+  - Resolve the addon version for the Cluster’s Kubernetes version.
+  - Provide addon configuration starting from a set of default settings but allowing per Cluster customization.
+  - Wait for the addon to be actually running.
+- Some of the addon must be upgraded before/during/after cluster (and more specifically control plane) upgrade; also in this case this operation brings in some additional requirements, like e.g. to proper cleanup the previous version of the addon.
+- There should be support for out-of band upgrades, e.g. for CVE fix (not linked to the cluster lifecycle).
+- It should be possible to orchestrate pre/post installation steps for each addon, thus allowing to manage migrations/operations like:
+  - CPI migrating from in-tree to out-of-tree
+  - CoreDNS configuration migration from one version to another
+- Addon deletion should be orchestrated before Cluster deletion
+
+The current proposal starts to address part of this problem by designing how Cluster API can orchestrate an external tool for addon management; the external tool is expected to provide all the foundational capabilities for addon management like e.g. addon packages/repository, templating/configuration, management of the creation, upgrade and deletion workflow etc.
+
+Cluster API - and most specifically a new Cluster API-aware component described in this proposal - should orchestrate the above tool so that Cluster lifecycle and addon lifecycle are consistent.
+
+
+### Goals
+
+- To design a solution for orchestrating Cluster add-on
+- To leverage an existing package management tool, Helm in this case, for all the foundational capabilities of addon management like e.g. addon packages/repository, templating/configuration, addon creation, upgrade and deletion etc. 
+
+### Non goals
+
+- To implement a full fledged package management tool in Cluster API; there are already several awesome package management tools in the ecosystem, and CAPI should not reinvent the wheel. 
+- To provide mechanism for altering, customizing or dealing with single Kubernetes resources defining a Cluster add-on (Deployments, Services, ServiceAccounts etc); CAPI should treat add-ons as opaque components and delegate all the operations impacting add-on internals to the package management tool.
+
+### Future work
+
+- Provide support for advanced orchestration use cases leveraging on Cluster API’s lifecycle hooks for Cluster API.
+- Introduce capabilities to coordinate operation accordingly to Add-on dependencies.
+- To eventually reconsider the future of ClusterResourceSet experimental features. 
+
+## Proposal
+
+This document introduces the concept of a Cluster API addon provider using [Helm](https://helm.sh/), a new, Kubernetes CRD called HelmChartProxy to be deployed on the Management Cluster. HelmChartProxy will act as an intermediary (proxy) between Cluster API and the Helm to install packages onto workload clusters.
+
+### User Stories
```

Comment:
This user stories should probably not be Helm specific and should cover all addon providers